File: personal-assistant-agent/src/tools/GetUnreadEmails.ts
import { google } from 'googleapis';
import type { OAuth2Client } from 'google-auth-library';

export async function getUnreadEmails(auth: OAuth2Client, maxResults: number): Promise<any[]> {
    const gmail = google.gmail({ version: 'v1', auth });

    const response = await gmail.users.messages.list({
        userId: 'me',
        q: 'is:unread',
        maxResults,
    });

    const messages = response.data.messages || [];
    const emailPromises = messages
        .filter((m): m is { id: string } => !!m && !!m.id)
        .map(async (message) => {
            const emailResponse = await gmail.users.messages.get({
                userId: 'me',
                id: message.id,
            });

            const fullDetails = emailResponse.data;
            const headers = fullDetails.payload?.headers || [];

            // Helper function to extract specific information from Google's header array
            const getHeader = (name: string) =>
                headers.find(h => h.name?.toLowerCase() === name.toLowerCase())?.value || 'Unknown';

            return {
                id: fullDetails.id,
                from: getHeader('From'),
                subject: getHeader('Subject'),
                date: getHeader('Date'),
                snippet: fullDetails.snippet,
            };
        });

    return Promise.all(emailPromises);
}

File: personal-assistant-agent/src/tools/get_unread_emails.py
from typing import Any, List

from google.oauth2.credentials import Credentials
from googleapiclient.discovery import build


def get_unread_emails(auth: Credentials, max_results: int) -> List[dict[str, Any]]:
    gmail = build("gmail", "v1", credentials=auth)

    response = (
        gmail.users()
        .messages()
        .list(userId="me", q="is:unread", maxResults=max_results)
        .execute()
    )

    messages = response.get("messages") or []
    emails: List[dict[str, Any]] = []

    for message in messages:
        if not message or not message.get("id"):
            continue
        email_response = (
            gmail.users()
            .messages()
            .get(userId="me", id=message["id"])
            .execute()
        )
        headers = (email_response.get("payload") or {}).get("headers") or []

        def get_header(name: str) -> str:
            for h in headers:
                if (h.get("name") or "").lower() == name.lower():
                    return h.get("value") or "Unknown"
            return "Unknown"

        emails.append(
            {
                "id": email_response.get("id"),
                "from": get_header("From"),
                "subject": get_header("Subject"),
                "date": get_header("Date"),
                "snippet": email_response.get("snippet"),
            }
        )

    return emails
